tools/testing/test_selections.py

```diff
@@ -67,97 +67,182 @@ def get_with_pytest_shard(
+    included = test.included()
+    excluded = test.excluded()
+    included_classes_duration = get_duration_for_classes(test.test_file, included)
+    excluded_classes_duration = get_duration_for_classes(test.test_file, excluded)
 
-        if duration and duration > THRESHOLD:
-            num_shards = math.ceil(duration / THRESHOLD)
-            for i in range(num_shards):
-                sharded_tests.append(
-                    ShardedTest(test, i + 1, num_shards, duration / num_shards)
-                )
-        else:
-            sharded_tests.append(ShardedTest(test, 1, 1, duration))
-    return sharded_tests
+    if included_classes_duration is None or excluded_classes_duration is None:
+        # Didn't get the time for all classes, so time is unknown
+        return None
 
+    if included:
+        return included_classes_duration
+    assert (
+        excluded
+    ), f"TestRun {test} is not full file but doesn't have included or excluded classes"
+    if file_duration is None:
+        return None
+    return file_duration - excluded_classes_duration
 
-def calculate_shards(
-    num_shards: int,
+
+def shard(
+    sharded_jobs: List[ShardJob],
     tests: Sequence[TestRun],
     test_file_times: Dict[str, float],
-    test_class_times: Optional[Dict[str, Dict[str, float]]],
-    must_serial: Optional[Callable[[str], bool]] = None,
+    test_class_times: Dict[str, Dict[str, float]],
+    estimated_time_limit: Optional[float] = None,
     sort_by_time: bool = True,
-) -> List[Tuple[float, List[ShardedTest]]]:
-    must_serial = must_serial or (lambda x: True)
-
-    known_tests: Sequence[TestRun] = tests
-    unknown_tests: Sequence[TestRun] = []
-
+    serial: bool = False,
+) -> None:
+    if len(sharded_jobs) == 0:
+        assert len(tests) == 0, "No shards provided but there are tests to shard"
+        return
+    # Modifies sharded_jobs in place
+    known_tests = tests
+    unknown_tests = []
     if sort_by_time:
         known_tests = [
             x
             for x in tests
-            if x.test_file in test_file_times
-            or (test_class_times and x.test_file in test_class_times)
+            if get_duration(x, test_file_times, test_class_times) is not None
         ]
         unknown_tests = [x for x in tests if x not in known_tests]
 
+        assert (
+            unknown_tests == [] or serial
+        ), f"Attmempting to parallelize unknown tests {unknown_tests}"
+    del tests
+
     known_tests = get_with_pytest_shard(known_tests, test_file_times, test_class_times)
 
     if sort_by_time:
         known_tests = sorted(known_tests, key=lambda j: j.get_time(), reverse=True)
 
-    sharded_jobs: List[ShardJob] = [ShardJob() for _ in range(num_shards)]
-    for test in known_tests:
-        if must_serial(test.name):
-            min_sharded_job = min(sharded_jobs, key=lambda j: j.get_total_time())
+    def _shard_serial(tests: List[ShardedTest], sharded_jobs: List[ShardJob]) -> None:
```

Comment:
Nits:  Do you think it's useful to add boolean field in ShardedTest to indicate if it's a serial or parallel shard?  That could be print to the log as part of the shard name for debugging?

Reply:
if a test gets run on the shard, it'll be printed out later whether it's serial or parallel, and I'd rather not clog up the logs with all this information